hasher-matcher-actioner/hmalib/common/models/signal.py

```diff
@@ -1,45 +1,96 @@
 # Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
 
+r"""Signal Models
+
+The structure of a signal and the data we want to retain on it is
+super-dependent on the source of the signal. Trying to make a "base" signal
+model can be futile because of the disparity of attributes and actions that
+different sources can have.
+
+Instead of abstracting out, for signals, we are creating separate models for
+each source with their own feature evolution. This reality must also manifest on
+the UI. eg. The actions you can take for threatexchange signals are distinct
+from the actions you can take on your local bank's signals.
+
+But, since the id-universes of the signals are distinct, how do we precisely
+refer to a signal when using a signal id? This is a little complex because a
+signal is uniquely defined by two of its attributes: `signal_source` and
+`signal_id`.
+
+So, every signal class needs to define a property `signal_source` which will be
+used in indexing, queries etc.
+
+When we add more signal sources, we'll understand the complexity and identify a
+clean path for supporting multiple signal-sources. For now, only support
+threatexchange.
+"""
+
 import datetime
 from enum import Enum
 import typing as t
-
 from dataclasses import dataclass, field, asdict
+
 from mypy_boto3_dynamodb.service_resource import Table
 from boto3.dynamodb.conditions import Attr, Key, And
 from botocore.exceptions import ClientError
+from threatexchange.content_type.meta import get_signal_types_by_name
+
+from threatexchange.signal_type.signal_base import SignalType
 
 from hmalib.common.models.models_base import DynamoDBItem
 
 
-class PendingOpinionChange(Enum):
+class PendingThreatExchangeOpinionChange(Enum):
     MARK_TRUE_POSITIVE = "mark_true_positive"
     MARK_FALSE_POSITIVE = "mark_false_positive"
     REMOVE_OPINION = "remove_opinion"
     NONE = "none"
 
 
 @dataclass
-class SignalMetadataBase(DynamoDBItem):
+class ThreatExchangeSignalMetadata(DynamoDBItem):
     """
-    Base for signal metadata.
-    'ds' refers to dataset which for the time being is
-    quivalent to collab or privacy group (and in the long term could map to bank)
+    This object is designed to be an ~lookaside on some of the values used by
+    MatchRecord for easier and more consistent updating by the syncer and UI.
+
+    We only write these objects when we match against a signal from
+    threatexchange. Updates can happen when:
+      a. a user registers an opinion against this signal.
+      b. the signal's attributes are updated in threatexchange.
+
+    User's registration of an opinion is a multi-stage process. The opinion is
+    first recorded on this object as a 'pending' opinion directly by the UI.
+    Once the opinion is written to threatexchange and synced back, the 'pending'
+    opinion is cleared.
+
+    Storage
+    ---
+    PK: s#{source_short_code}#{signal_id}
+    SK: #threatexchange-signal-metadata
```

Comment:
Unfortunately SK as ds_id (dataset i.e. privacy_group_id) is currently used as to determine uniqueness so I don't think this change will work. Because ThreatExchangeOpinionChange has that level of the granularity: e.g. users could agree this signal belongs in the `cat` dataset but not the `dog` dataset.

Reply:
Aah! Because the same indicator can be in multiple privacy groups? Gotcha!

Well, back to the drawing board. Thanks for the flag!